# Spam Detector

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [10]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head(25)

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0             0.00               0.64           0.64           0.0   
1             0.21               0.28           0.50           0.0   
2             0.06               0.00           0.71           0.0   
3             0.00               0.00           0.00           0.0   
4             0.00               0.00           0.00           0.0   
5             0.00               0.00           0.00           0.0   
6             0.00               0.00           0.00           0.0   
7             0.00               0.00           0.00           0.0   
8             0.15               0.00           0.46           0.0   
9             0.06               0.12           0.77           0.0   
10            0.00               0.00           0.00           0.0   
11            0.00               0.00           0.25           0.0   
12            0.00               0.69           0.34           0.0   
13            0.00               0.00           0.00           0.0   
14            0.00               0.00           1.42           0.0   
15            0.00               0.42           0.42           0.0   
16            0.00               0.00           0.00           0.0   
17            0.00               0.00           0.00           0.0   
18            0.00               0.00           0.55           0.0   
19            0.00               0.63           0.00           0.0   
20            0.00               0.00           0.00           0.0   
21            0.05               0.07           0.10           0.0   
22            0.00               0.00           0.00           0.0   
23            0.00               0.00           0.00           0.0   
24            0.00               0.00           0.00           0.0   

    word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0            0.32            0.00              0.00                0.00   
1            0.14            0.28              0.21                0.07   
2            1.23            0.19              0.19                0.12   
3            0.63            0.00              0.31                0.63   
4            0.63            0.00              0.31                0.63   
5            1.85            0.00              0.00                1.85   
6            1.92            0.00              0.00                0.00   
7            1.88            0.00              0.00                1.88   
8            0.61            0.00              0.30                0.00   
9            0.19            0.32              0.38                0.00   
10           0.00            0.00              0.96                0.00   
11           0.38            0.25              0.25                0.00   
12           0.34            0.00              0.00                0.00   
13           0.90            0.00              0.90                0.00   
14           0.71            0.35              0.00                0.35   
15           1.27            0.00              0.42                0.00   
16           0.94            0.00              0.00                0.00   
17           0.00            0.00              0.00                0.00   
18           1.11            0.00              0.18                0.00   
19           1.59            0.31              0.00                0.00   
20           0.00            0.00              0.00                0.00   
21           0.76            0.05              0.15                0.02   
22           2.94            0.00              0.00                0.00   
23           1.16            0.00              0.00                0.00   
24           0.00            0.00              0.00                0.00   

    word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0              0.00            0.00  ...        0.000        0.000   
1              0.00            0.94  ...        0.000        0.132   
2              0.64   

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I think that random forests will do better because spam data is very non-linear.

## Split the Data into Training and Testing Sets

In [15]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
y.head()

0    1
1    1
2    1
3    1
4    1
Name: spam, dtype: int64

In [21]:
X = data.copy()
X = X.drop(columns='spam')
X.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0         0.00   
1             0.00            0.94  ...                   0.0         0.00   
2             0.64            0.25  ...                   0.0         0.01   
3             0.31            0.63  ...                   0.0         0.00   
4             0.31            0.63  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.000          0.0        0.778        0.000        0.000   
1        0.132          0.0        0.372        0.180        0.048   
2        0.143          0.0        0.276        0.184        0.010   
3        0.137          0.0        0.137        0.000        0.000   
4        0.135          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  
0                       278  
1                      1028  
2                      2259  
3                       191  
4                       191  

[5 rows x 57 columns]

In [27]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [29]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=10)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [33]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)

In [47]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_train_scaled

array([[-0.34103614, -0.1663159 ,  0.17225374, ..., -0.09501736,
        -0.25419538, -0.23289893],
       [-0.34103614, -0.1663159 , -0.54716092, ..., -0.11710528,
        -0.35556885, -0.4253799 ],
       [-0.34103614, -0.1663159 ,  0.61945745, ..., -0.10526536,
        -0.27758926, -0.40111759],
       ...,
       [-0.34103614, -0.1663159 , -0.15828813, ..., -0.12748594,
        -0.3087811 , -0.27980606],
       [-0.05261286, -0.1663159 , -0.02218266, ..., -0.04450699,
         0.14350053,  0.40115604],
       [ 0.90879804, -0.1663159 , -0.54716092, ..., -0.05077518,
        -0.16061987, -0.28304103]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [53]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression().fit(X_train_scaled, y_train)
print(f'Model: {type(lr).__name__}')
print(f'Train score: {lr.score(X_train_scaled, y_train)}')
print(f'Test Score: {lr.score(X_test_scaled, y_test)}')

Model: LogisticRegression
Train score: 0.9246376811594202
Test Score: 0.9296264118158123


In [57]:
# Make and save testing predictions with the saved logistic regression model using the test data
logistic_predictions = lr.predict(X_test_scaled)

# Review the predictions
print(logistic_predictions)

[1 1 0 ... 0 0 0]


In [59]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, logistic_predictions)
print(accuracy)

0.9296264118158123


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [64]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier().fit(X_train_scaled, y_train)
print(f'Model: {type(rf_model).__name__}')
print(f'Train score: {rf_model.score(X_train_scaled, y_train)}')
print(f'Test Score: {rf_model.score(X_test_scaled, y_test)}')

Model: RandomForestClassifier
Train score: 0.9994202898550725
Test Score: 0.9609035621198957


In [68]:
# Make and save testing predictions with the saved logistic regression model using the test data
forest_predictions = rf_model.predict(X_test_scaled) 

# Review the predictions
print(forest_predictions)

[1 1 0 ... 0 0 1]


In [72]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, forest_predictions)
accuracy

0.9609035621198957

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*LogisticRegression*
Train score: 0.9246376811594202
Test Score: 0.929626411815812

*RandomForestClassifier*
Train score: 0.9994202898550725
Test Score: 0.960903562119895

This makes sense with my prediction due to the non-linearity of spam. The data has very non-linear spam patterns. I do also think that random forests overfit a lot though, so it might be better to use a less complex model.73